lua/ttt2/libraries/thermalvision.lua

```diff
@@ -0,0 +1,263 @@
+
+			--weaken the effect behind walls
+			cam.Start2D()
+				local overlayColor = bgColoring and overlayColorWithBG or overlayColorWithoutBG
+				surface.SetDrawColor(overlayColor)
+				surface.DrawRect(0, 0, ScrW(), ScrH())
+			cam.End2D()
+		end
+
+		--draw the blue colored background
+		if bgColoring then
+			render.SetStencilCompareFunction(STENCIL_NOTEQUAL)
+			render.SetStencilZFailOperation(STENCIL_KEEP)
+			render.SetStencilPassOperation(STENCIL_KEEP)
+			render.SetStencilFailOperation(STENCIL_KEEP)
+
+			DrawColorModify(colorModifyBG)
+		end
+
+		--second render pass: draw all parts of entities which are visible (z test succceeds)
+		if #thermalvisionList > 0 then
+			render.ClearStencil()
+
+			--draw stencil mask only for visible parts
+			render.SetStencilCompareFunction(STENCIL_ALWAYS)
+			render.SetStencilZFailOperation(STENCIL_KEEP)
+			render.SetStencilPassOperation(STENCIL_REPLACE)
+			render.SetStencilFailOperation(STENCIL_ZERO)
+
+			--render all entities but keep depth testing
+			render.SuppressEngineLighting(true)
+			for i = 1, #thermalvisionList do
+				local entry = thermalvisionList[i]
+				local ent = entry.ent
+				local mode = entry.mode
+
+				if not IsValid(ent) then continue end
+
+				--overwrite with the original model and prevent effect appliance afterwards if the effect should be only applied for non visible things
+				if mode == THERMALVISION_MODE_NOTVISIBLE then
+					render.SetStencilWriteMask(0)
+					render.SuppressEngineLighting(false)
+				end
+
+				ent:DrawModel()
+
+				render.SetStencilWriteMask(255)
+				render.SuppressEngineLighting(true)
+			end
+			render.SuppressEngineLighting(false)
+
+			--screenspace effect on stenciled areas again
+			render.SetStencilCompareFunction(STENCIL_EQUAL)
+			render.SetStencilZFailOperation(STENCIL_KEEP)
+			render.SetStencilPassOperation(STENCIL_KEEP)
+			render.SetStencilFailOperation(STENCIL_KEEP)
+
+			DrawColorModify(colorModifyEnts)
+		end
+
+		render.SetStencilEnable(false)
+	end
+
+	---
+	-- Hook adding
+	-- @realm client
+	local function AddThermalvisionHook()
+		if thermalvisionHookInstalled then return end
+
+		hook.Add("PostDrawTranslucentRenderables", "RenderThermalvision", RenderHook)
+
+		thermalvisionHookInstalled = true
+	end
+
+	---
+	-- Hook removing
+	-- @realm client
+	local function RemoveThermalvisionHook()
+		hook.Remove("PostDrawTranslucentRenderables", "RenderThermalvision")
+
+		thermalvisionHookInstalled = false
+	end
+
+	---
+	-- Clearing the cached @{Entity} list
+	-- @realm client
+	function thermalvision.Clear()
+		thermalvisionList = {}
+
+		RemoveThermalvisionHook()
+	end
+
+	---
+	-- internal entity removing
+	-- @realm client
+	local function RemoveInternal(ents)
+		local thermalvisionListSize = #thermalvisionList
+		if thermalvisionListSize == 0 or #ents == 0 then return end
+
+		for i = 1, #ents do
+			for j = 1, #thermalvisionList do
+				if thermalvisionList[j].ent == ents[i] then
+					--for now only setting it to nil
```

Comment:
space after inline comment. If there is no space, it's commented code by default